src/crimson/net/Workspace.h

```diff
@@ -0,0 +1,59 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel Corp.
+ *
+ * Author: Yingxin Cheng <yingxincheng@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include "Protocols.h"
+
+namespace ceph::net {
+
+class Workspace
+{
+  NoneProtocol none_protocol;
```

Comment:
why do we need to keep all the states around?

Reply:
It has a similar idea with:
https://github.com/ceph/ceph/blob/4470842ebf2e2a464d030cca24616d306442e9dc/src/msg/async/Protocol.h#L46-L49

`Workspace` pre-allocates memory space needed for the state machine, plus it doesn't need to request memory from heap multiple times for every concrete `Protocol` class.